### PR TITLE
Add tests for rest controllers

### PR DESCRIPTION
**Pull request by randrii**

### CODE REVIEW

File: src/test/java/com/rybka/controller/BankControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+import static org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup;
+
+@WebMvcTest(BankController.class)
+class BankControllerTest {
+
+    private MockMvc mockMvc;
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/BankControllerTest.java:27]: Try to place Autowire anotation and remove builder in the setUp method

Comment (jarc0der) [reply src/test/java/com/rybka/controller/BankControllerTest.java:27]: and remove controller from there

File: src/test/java/com/rybka/controller/BankControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    @BeforeEach
+    void setUp() {
+        var controller = new BankController(service);
+        mockMvc = standaloneSetup(controller).build();
+    }
+
+    @SneakyThrows
+    @Test
+    public void testOnProperGetResponse() {
+        var currencyBase = "usd";
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/BankControllerTest.java:37]: //given 

//when

//then

or 

//given

//expects

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    @SneakyThrows
+    @Test
+    public void testOnProperPostResponse() {
+        var base = "USD";
+        var type = "BUY";
+        var count = 4d;
+        var targetUrl = "/exchange";
+        var exchangeResultData = new ExchangeResultData(base, count, List.of());
+        var exchangeRequest = new ExchangeRequest(base, type, count);
+        when(service.retrieveTopCurrency(anyString(), anyString(), anyDouble())).thenReturn(exchangeResultData);
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/ExchangeControllerTest.java]: why anyString() anyDouble()? We have that values and can pass them

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        var exchangeResponseDataJson = mapToJson(exchangeResultData);
+        var exchangeRequestJson = mapToJson(exchangeRequest);
+
+        mockMvc.perform(post(targetUrl)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(exchangeRequestJson))
+                .andDo(print())
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/ExchangeControllerTest.java]: no need for print()

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                .andExpect(status().isOk())
+                .andExpect(content().string(exchangeResponseDataJson));
+    }
+
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/ExchangeControllerTest.java:79]: What if exchangeRequest is broken json? We should test that.

File: src/test/java/com/rybka/controller/HistoryControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        var currencyTarget = currencyList.get(random.nextInt(currencyList.size()));
+        var currencyRate = random.nextDouble();
+        var currencyAmount = random.nextInt();
+        var total = currencyRate * currencyAmount;
+
+        return new CurrencyHistory(currencyBase, currencyTarget, currencyAmount, currencyRate, total);
+    }
+
+    @SneakyThrows
+    private String mapToJson(List<CurrencyHistory> histories) {
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/HistoryControllerTest.java]: Create class with Utils to utilize current method to reduce code duplication. 
This method is used in all tests. ExchangeControllerTest has generic method that can be moved and utilized.

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                .andExpect(content().string(currencyDataJson));
+    }
+
+    @SneakyThrows
+    @Test
+    public void testOnProperPostResponse() {
+        var base = "USD";
+        var type = "BUY";
+        var count = 4d;
+        var targetUrl = "/exchange";
```
Comment (jarc0der) [inline src/test/java/com/rybka/controller/ExchangeControllerTest.java]: Would be better to have constants class for endpoints and utilize them in the code and tests.

File: src/main/java/com/rybka/model/ExchangeRequest.java [hunk 8]
```diff
@@ -2,9 +2,11 @@
 
 import lombok.AllArgsConstructor;
 import lombok.Data;
+import lombok.NoArgsConstructor;
 
 @Data
 @AllArgsConstructor
+@NoArgsConstructor
```
Comment (jarc0der) [inline src/main/java/com/rybka/model/ExchangeRequest.java:9]: Bad idea to change production code for tests :(

File: src/test/java/com/rybka/controller/BankControllerTest.java (hunk 1 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/BankControllerTest.java:27]: fixed

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java (hunk 3 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/ExchangeControllerTest.java]: fixed

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java (hunk 4 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/ExchangeControllerTest.java]: fixed

File: src/test/java/com/rybka/controller/HistoryControllerTest.java (hunk 6 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/HistoryControllerTest.java]: fixed

File: src/main/java/com/rybka/model/ExchangeRequest.java (hunk 8 above)
Comment (randrii) [reply src/main/java/com/rybka/model/ExchangeRequest.java:9]: fixed

File: src/test/java/com/rybka/controller/BankControllerTest.java (hunk 2 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/BankControllerTest.java:37]: fixed

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java (hunk 7 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/ExchangeControllerTest.java]: fixed

File: src/test/java/com/rybka/controller/ExchangeControllerTest.java (hunk 5 above)
Comment (randrii) [reply src/test/java/com/rybka/controller/ExchangeControllerTest.java:79]: fixed

**Pull request merged**